### PR TITLE
React-i18next Generic translate

### LINKED ISSUE #16395: react-i18next typings should use generics for proper typing
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @KostyaEsmukov

Right now, the typings require you to do

```
import { InjectedTranslateProps } from 'react-i18next'

interface MyComponentProps extends InjectedTranslateProps {
  ...
}
```

However, this results in unnecessary behavior with the `t` function, because it's marked as optional (to avoid causing errors in other components when importing a translated component). The `t` function always exists, it's just not actually an externally available prop.

Instead, the extend mentioned above shouldn't be needed, and the `translate` HOC function should work something like:

```
import { TranslateProps, translate } from 'react-i18next'

const MyComponentImpl: React.StatelessComponent<MyComponentProps & TranslateProps>;

const MyComponent: React.StatelessComponent<MyComponentProps> = translate<MyComponentProps>(MyComponentImpl);
```

This could use typings something like (very briefly giving an idea):

```
interface TranslateProps {
  t: (text: string) => string
}

function translate<P, S>(component: React.Component<P & TranslateProps, S>): React.Component<P, S>
function translate<P>(component: React.StatelessComponent<P & TranslateProps>): React.StatelessComponent<P>
```

The idea would be for it work very similarly to the typings for react-router or react-redux, which inject the appropriate props while still exporting the component with only its own props.

### PR DESCRIPTION
**Pull request by MichaReiser**
Fixes #16395

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [translate.js](https://github.com/i18next/react-i18next/blob/master/src/translate.js)
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Warning: 

This is a breaking change. Like for react-redux, using the HOC as a decorator is not supported (#9951). I don't know of any reasonable solution for this limitation. The current implementation provides better typing support as it does not specify `t` and `i18n` as optional, instead these properties are mandatory but set by the translate HOC. The previous solution worked because it just returned the type of the Component which is less accurate than what this solution provides. 

Please advise how to continue in regards to the breaking change.

### CODE REVIEW

**Comment by ikappas** [comment]
@MichaReiser The I18n react component is missing from this.

**Comment by MichaReiser** [comment]
@ikappas Can you elaborate what you mean with missing from this?

**Comment by ikappas** [comment]
@MichaReiser

Looking at the react-i18next sources I see a `18n` ReactComponent that is exported from here: 
https://github.com/i18next/react-i18next/blob/master/src/I18n.js 

In this PR I don't see this component being defined anywhere.

Should't there be a definition for `I18n` something like so:
```js
  export interface I18nProps extends Omit<TranslateOptions, 'translateFuncName'> {
    ns?: string;
  }

  export class I18n extends React.Component<I18nProps> { }
```

**Comment by MichaReiser** [comment]
@ikappas Thanks for your explanations. 

I think the usage of I18n to realise the HOC is an implementation detail and, therefore, should not be expressed in the typings. However, the [documentation](https://react.i18next.com/components/translate-hoc.html) mentions some properties that I have added (and was not aware of). 

The Typings for the context are intentionally limited to what is needed in this PR. The full context properties are added in #21881.

File: types/react-i18next/index.d.ts [hunk 1]
```diff
@@ -1,10 +1,10 @@
-// Type definitions for react-i18next 4.6
+// Type definitions for react-i18next 7.0
```
Comment (minestarks) [inline types/react-i18next/index.d.ts]: Could you back up the old version to a folder called v4 as described in the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-update-a-package-to-a-new-major-version)? This should at least mitigate the breaking change.

**Review by minestarks — CHANGES_REQUESTED** [review]

File: types/react-i18next/index.d.ts (hunk 1 above)
Comment (minestarks) [reply types/react-i18next/index.d.ts]: Actually looks like PR #21881 is also updating the version. Best not duplicate effort.

Comment (MichaReiser) [reply types/react-i18next/index.d.ts]: Fixed

**Review by minestarks — APPROVED** [review]

**Comment by minestarks** [comment]
Thanks - one final thing, it looks like the bot failed to find and ping original author, so I want to give them a chance to comment

@GiedriusGrabauskas would you like to review?

**Comment by rosskevin** [comment]
The I18n component is not an implementation detail, it is one of two primary ways of interfacing with i18next https://react.i18next.com/components/i18n-render-prop.html

**Comment by MichaReiser** [comment]
Thx for merging.

I totally agree with you. The I18n component itself is not an implementation detail. However, since the documentation doesnt state that translate is implemented using the i18n component I believe it is the best to also not expose this information in the typings of translate. I believe there is another PR that adds the typings for the component.

**Pull request merged**